### PR TITLE
feat(Smrnmi): add support to raise non-maskable interrupt

### PR DESCRIPTION
**Pull request by lewislzh**

### CODE REVIEW

File: src/test/csrc/difftest/refproxy.h [hunk 1]
```diff
@@ -93,6 +93,11 @@ class RefProxyConfig {
   bool enable_store_log = false;
 };
 
+#ifdef CPU_XIANGSHAN
```
Comment (poemonsense) [inline src/test/csrc/difftest/refproxy.h:96]: The `raise_nmi_intr` should be declared as an optional interface. See `#define REF_OPTIONAL(f)  ...`

**Review by poemonsense — COMMENTED** [review]
Will NMI interrupt and normal interrupt occur at the same time?

Generally speaking, NMI should be implemented as an optional feature in DiffTest. For REFs with nmi intr, they should implement the raise_nmi_intr interface. Otherwise, if they do not implement it, NMI will be skipped.

Therefore, probably you should refer to the implementation of guided_exec (https://github.com/OpenXiangShan/difftest/blob/master/src/test/csrc/difftest/refproxy.h#L228-L230) and follow similar ideas to implement `Difftest::do_interrupt`. For example, check if the raise_nmi_intr exists, and then decide whether to call it. Using `CPU_XIANGSHAN` to check the existence of nmi_intr is a bad idea.

File: src/test/csrc/difftest/difftest.cpp [hunk 2]
```diff
@@ -398,6 +398,9 @@ int Difftest::step() {
 
 void Difftest::do_interrupt() {
   state->record_interrupt(dut->event.exceptionPC, dut->event.exceptionInst, dut->event.interrupt);
+#ifdef CPU_XIANGSHAN
+  proxy->raise_nmi_intr(dut->event.hasNMI);
```
Comment (poemonsense) [inline src/test/csrc/difftest/difftest.cpp:402]: Something like:

```c++
if (proxy->raise_nmi_intr) {
  // call it
}
else {
  // raise some warnings and exit or simply leave it empty
}
```

Comment (lewislzh) [reply src/test/csrc/difftest/difftest.cpp]: I have changed nmi as an optional feature like guided_exec.

Comment (lewislzh) [reply src/test/csrc/difftest/difftest.cpp]: NMI interrupt and normal interrupt can occur at the same time, and NMI interrupt has the highest priority.

File: src/test/csrc/difftest/refproxy.h [hunk 3]
```diff
@@ -225,6 +226,12 @@ class RefProxy : public AbstractRefProxy {
     }
   }
 
+  inline void trigger_nmi(bool hasNMI) {
+    if (raise_nmi_intr) {
+      raise_nmi_intr(hasNMI);
+    }
```
Comment (poemonsense) [inline src/test/csrc/difftest/refproxy.h:232]: Adding an else and saying raise_nmi_intr has not been supported by the REF?

**Review by poemonsense — APPROVED** [review]
Look good to me.

However, it may be better to add the else for debugging usages. Otherwise, if some REF does not support NMI, it would be difficult for the user to know that it is probably due to the missing definition in the REF.

**Pull request merged**